### ISSUE TITLE
SoS Running Modes

### ISSUE THREAD

**Issue by BoPeng** [BoPeng/SOS#230]
There are a number of problems with evaluating statements in prepare mode. Other than performance problem caused by repeated evaluation and generation of output, it is difficult to decide how to handle errors. For example, if a step reads a file generated by a previous step, reading a non-existing file causes error and causing an outdated file can cause even bigger problem.

A proper solution, I think, is to perform deep but static analysis of workflow and employ a real dynamic approach in execution. I mean, we can now obtain variables used by a step through static python parser. We can then build very basic DAG with most input and output as `undetermined` other than simple cases such as no input, empty input etc. The DAG would then be updated during workflow execution.

Overall this would be a difficult but safer and more efficient approach.

**Comment by gaow** [comment]
Move `dryrun` related discussion here. What else does `dryrun` do other than evaluating Python statements while running a "dry" SoS script?

**Comment by BoPeng** [comment]
`dryrun` should not download stuff, for example, and I think we should move the transcript stuff here. Basically dryrun is a way, to the best of SoS's power without executing statements, to check what steps are involved, what commands would be executed etc... However, `check_command` sounds like something dryrun could do... 

On the `run` side, because of recent improvement in parsing steps, build DAG and submit jobs, we can execute large  workflows in parallel without running the workflow in prepare mode, which is very problematic as far as I can see.

**Comment by BoPeng** [comment]
Actually executing statements in `dryrun` should be ok because it does not slow down `run`. This would give much more accurate acount on what the workflow does.

**Comment by gaow** [comment]
I thought `check_command` is job of `prepare` ... To me, `prepare` is a way to check resource requirement without executing statements. The difference between `dryrun` and `run` should be whether or not to run any actual scripts -- this is what we'd expect in other software, that `dryrun` is like echo commands to be executed without actually executing it.

Do we always have to save transcripts, though? Transcript is linear and does not directly reflect the execution logic. I do not find it is very useful.

It might be helpful to recap the roles of inspect, prepare and dryrun here (that you can edit)

mode | function
--------------:|:-----------------
`inspect` | Check SoS syntax 
`prepare` | Check required resources, e.g., command tools, docker images, and handle downloads
`dryrun` | Check steps involved. Python statements will be executed in this mode

**Comment by BoPeng** [comment]
The point here is that I do not want to run `prepare` in an incomplete and potentially wrong way, and execute all statements again for `run` mode. Basically I would like to separate `inspect`, `prepare`, and `run` mode to avoid repeated execution but I do not mind having them each do part of the work, for example

mode| function
--------------:|:-----------------
`inspect` | Check SoS syntax without executing anything. This step will be done by any other mode and is not really needed.
`dryrun` | Execute all statements but do not execute action. Probably output transcripts. This might not work well because some statements migt not work without execution of actions. 
`prepare` | Execute all statements and actions in prepare mode. Check required resources, e.g., command tools, docker images, and handle downloads
`run` | Execute all statements and all actions.

The `run` mode will not be based on dag built from `prepare` mode because that DAG can be wrong (e.g. reading an obsolete file that should have been freshly generated).

We can do many things and try to be clever when prepare fails but I now feel that the most candid solution is to stop when prepare cannot proceed, and tells user that the error might be caused by previous steps that have not been executed completely.

**Comment by gaow** [comment]
So the default mode is `prepare + run`, and we have `dryrun` with an additional option of whether or not to export transcript? For the interface, `sos run` will only have `-i` or `-p` options, and we make `dryrun` a separate command?

**Comment by BoPeng** [comment]
No, `prepare`, if remains, will be separated from `run` so that `run` can be executed faster without having to execute statements multiple times. The challenge is then how to update DAG during execution.

**Comment by BoPeng** [comment]
OK, `sos run` has been implemented to be independent of `prepare`. It now uses a real dynamic DAG and should be free of `prepare` caused problems we had before.

It is then time to decide what `inspect`, `dryrun`, and `prepare` should do. It can be overkilling with all these options so I am ok with letting go some or even all of them.

**Comment by BoPeng** [comment]
Since all running modes are separated, I propose that we have the following modes:

mode| function
--------------:|:-----------------
`dryrun` | Execute all statements but do not execute action.  
`run` | Execute all statements and all actions.

We currently have `inspect` but the `dryrun` sounds more appropriate (because we actually execute statement) and should probably be used. 

`prepare` is a good idea but having this mode significantly complicates the design of SoS so I am not sure if the benefit worth the trouble. Removing prepare will greatly simplify the design of actions because it has only two modes: dryrun (not run) or run. Another point is that if users really want to have a prepare mode, they can create a sub-workflow as follows:

```
[download_resource, prepare_0: provides='resource']

[step_100, prepare_10]

```

Then run 

```
sos run script prepare
```

to execute resource preparation steps.

**Comment by gaow** [comment]
I agree with removing `inspect`. `prepare` is useful, but if it complicates the design very much I'd agree removing it as well. I have mixed feelings in using sub-workflow for resource preparation. But I think it's more an issue of style and personal preference.

How do we deal with transcripts? I think it makes sense to have it in `dryrun` not in `run`, but that will involve evaluating action strings in `dryrun`. Should we make it optional?

**Comment by BoPeng** [comment]
`prepare` complicates the design but the feature is already there and largely usable ... so it is a tough decision if we should remove it. I agree that subworkflow is not that easy to use. 

The bottom line is that we should remove a feature if it does not work well. The prepare mode currently does not work well with my RNASeq workflow because some of the steps depends on results generated from previous steps, and `prepare` would fail. I would keep the prepare mode if it does not behave this way, and the solution might be #232, namely we only execute a step if it can be safely "prepared" so that any error would indeed be user error.

The same logic goes to transcript, if we cannot get reliable transcript in dryrun mode, perhaps we should do it in run mode.

**Comment by BoPeng** [comment]
There is another issue with prepare mode. Previous we run script in prepare and run mode, so some steps can be run in prepare mode and be skipped in run mode. Right now we only run in run mode so steps that are executable only in prepare mode would be ignored....

**Comment by BoPeng** [comment]
Because `run` no longer calls `prepare`, actions in `prepare` must also be called in `run` mode. So in the end we can say "a subset of actions can also be executed in prepare mode"...

**Comment by BoPeng** [comment]
We will have `--dryrun (-n)`, `--prepare (-p)` and `run` mode. The prepare mode can be removed if it does not work well.

**Comment by BoPeng** [comment]
Also https://github.com/BoPeng/SOS/commit/4fe66db92370b3645e085389c18340f371fd547f and https://github.com/BoPeng/SOS/commit/147ad94c17d8f6d8048f7abd48672292423bc911
